assignments/burmanm/week-1.scm

```diff
@@ -0,0 +1,82 @@
+; The Orgmode was so buggy that I had to revert to something else. No time to fight with it right now
+;; Exercise 1.1
+10
+12
+8
+3
+6
+a
+b
+19
+#f
+4
+16
+6
+16
+
+;; Exercise 1.2
+(/ (+ 5 4 (- 2 (- 3 (+ 6 (/ 4 5))))) (* 3 (- 2 7) (- 6 2)))
+
+;; Exercise 1.3
+;; My Scheme "leaflet/syntax notes" had the cond answer, so I wanted to use something a bit different
+(define (bigger a b) (if (> a b) a b))
+(define (square x) (* x x))
+(define (sum-squares x y) (+ (square x) (square y)))
+;; The following actually has 4 different output possibilities, with two permutations being the same (y x), (x y)
+(define (square-sum-larger x y z)
+  (if (> y x)
+      (sum-squares y (bigger x z))
+      (sum-squares x (bigger y z))))
+
+;; Exercise 1.4
+;; Since each expression can return a new expression, the code will always return a + |b| since negative b will use operand that creates a double negative operation.
+
+;; Exercise 1.5
+;; The difference is that applicative order would evaluate the subparts first, in which case
+;; the (p) would be the first to be expanded. That causes a never ending loop.
+;; With normal order, the (test 0 (p)) is first evaluated to:
+;; (if (= 0 0) 0 (p)) which is then executed and 0 is returned as (p) is never expanded.
+
+;; Exercise 1.6
+; The if is a special form function because it evaluates the "then" and "else" parts only after the value of the conditional is known. However, with the cond this isn't true and the "then" and "else" parts are evaluated first and only then the actual conditional. Alissa program will enter an indefinite loop since it will always evaluate sqrt-iter (which evaluates sqrt-iter, etc), even if the estimate is good enough.
+
+;; Exercise 1.7
+;; The error boundary for small numbers is too large, thus the difference of 0.001 is enough to cause
+;; significant error to the end result. For example, (sqrt 0.00001) outputs 0.0313564 when the real
+;; answer is close to 0.00316. The difference is already a magnitude.
+
+;; With large numbers, the algorithm suffers from lack of exactness, since changes will return the
+;; same value but this is beyond the allowed error rate. Any extra calculations will return the same
+;; number and thus the loop needes re-evaluation again and again.. For example:
+;; scheme@(guile-user) [1]> (sqrt 1234567890123)
+;; $17 = 1111111.1061109055
+;; scheme@(guile-user) [1]> (sqrt 12345678901234)
```

Comment:
That's just something I tried by hand without trying to find the correct number. But thanks for a great learning idea.. how to find the correct number with Scheme. 10^12 should still work, but it will fail somewhere before 10^13. Not the prettiest solution I guess (I don't know how to catch that never-ending-loop failure with Scheme):

```scheme
(define print-return
  (lambda (x)
    (display x)
    (newline)
    x))

;; tail-recursion so we prevent stack failures..
(define (find-largest start)
  (define (iter a)
    (if (sqrt a)
	(iter (print-return (+ a 1)))
	a))
  (iter start))

(find-largest  4398053520000)
```
I used quite a bit larger numbers for printing initially, such as:

```scheme
(define print-return
  (lambda (x)
    (if (= (remainder x 10000) 0)
	(begin
	  (display x)
	  (newline)
	  x)
	x)))
```

Smallest number I could find was ``4398053528663``

Reply:
To catch the never-ending-loop, you could check that the improved guess is equal to the last guess. Also, smallest number I found was: 4398053504138.8564453125 ;)